store/sqlstore/plugin_store.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package sqlstore
+
+import (
+	"database/sql"
+	"fmt"
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/store"
+)
+
+type SqlPluginStore struct {
+	SqlStore
+}
+
+func NewSqlPluginStore(sqlStore SqlStore) store.PluginStore {
+	s := &SqlPluginStore{sqlStore}
+
+	for _, db := range sqlStore.GetAllConns() {
+		table := db.AddTableWithName(model.PluginKeyValue{}, "PluginKeyValueStore").SetKeys(false, "PluginId", "Key")
+		table.ColMap("Value").SetMaxSize(8192)
+	}
+
+	return s
+}
+
+func (ps SqlPluginStore) CreateIndexesIfNotExists() {
+}
+
+func (ps SqlPluginStore) SaveOrUpdate(kv *model.PluginKeyValue) store.StoreChannel {
+	return store.Do(func(result *store.StoreResult) {
+		if result.Err = kv.IsValid(); result.Err != nil {
+			return
+		}
+
+		if ps.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+			// Unfortunately PostgreSQL pre-9.5 does not support upsert by default
+			// This wCTE will reduce the odds of a race but does not eliminate the chance
+			// of getting unique key violations
+			if _, err := ps.GetMaster().Exec(`
+				WITH new_values (PluginId, PKey, PValue) AS (VALUES (:PluginId, :Key, :Value)),
+				upsert AS (
+					UPDATE PluginKeyValueStore pkv SET PValue = :Value
+					FROM new_values nv
+					WHERE pkv.PluginId = nv.PluginId AND pkv.PKey = nv.PKey
+					RETURNING pkv.*
+				)
+				INSERT INTO PluginKeyValueStore (PluginId, PKey, PValue)
+				SELECT PluginId, PKey, PValue FROM new_values
+				WHERE NOT EXISTS (
+					SELECT 1 FROM upsert up WHERE up.PluginId = new_values.PluginId AND up.PKey = new_values.PKey)
+                `, map[string]interface{}{"PluginId": kv.PluginId, "Key": kv.Key, "Value": kv.Value}); err != nil {
```

Comment:
This is really gross. I wonder what the performance difference would be between this and just doing two queries:

```sql
INSERT INTO PluginKeyValueStore (PluginId, PKey, PValue) VALUES(:PluginId, :Key, :Value)
UPDATE PluginKeyValueStore SET PValue = :Value) WHERE PluginId = :PluginId AND PKey = :Key
```

If the insert fails due to anything but a unique key constraint, return the error without trying the update. If the update fails to modify any rows, that's okay; it just means that another client simultaneously issued a delete and pretending the operation was a success is perfectly valid behavior.

It would at least be portable. 🤷‍♂️

Reply:
Or better yet, I guess you could do this the other way around:

```sql
UPDATE PluginKeyValueStore SET PValue = :Value WHERE PluginId = :PluginId AND PKey = :Key
INSERT INTO PluginKeyValueStore (PluginId, PKey, PValue) VALUES(:PluginId, :Key, :Value)
```

If the update succeeds, we can return early. If the update doesn't affect any rows, try the insert. If the insert fails due to a unique key constraint, another client simultaneously issued another set, and again, pretending the operation was a success without actually setting the value is perfectly valid behavior.

Reply:
Yeah, it's definitely pretty gross. I'm not too worried about the performance of doing two separate queries, though it will make a bigger gap for a race to occur in.

I think whether or not updating or inserting is more common will depend on the plugin, but I'm leaning towards making the update the first query because inserts can only happen once for a key while there's potential for many updates on a key

Reply:
Ah, now I understand what the retries were for in `plugin.go`. If either of these approaches lets us remove the retry logic, that would be great, because the retry logic is pretty tightly coupled to this implementation.